providers/default_rhel.rb

```diff
@@ -0,0 +1,172 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+provides :kernel_module, platform_family: 'rhel'
+
+def whyrun_supported?
+  true
+end
+
+use_inline_resources
+
+action :save do
+  failing_sanity_check
+
+  # Requirements
+  package node['kernel_modules']['packages']
+
+  # Module init loading section
+  template modload_file do
+    source "#{node['platform_family']}-#{node['platform_version'].to_i}/init-load.erb"
+    cookbook 'kernel-modules'
+    mode '0755'
+    variables(
+      name: new_resource.name,
+      bin: modprobe_bin,
+    )
+  end
+
+  # Module loading options
+  resource_template_loading
+
+  resource_unloading(:save)
+
+  resource_load
+end
+
+action :load do
+  failing_sanity_check
+
+  # Module loading options
+  resource_template_loading(:load)
+
+  resource_unloading(:load)
+
+  resource_load(:load)
+end
+
+action :unload do
+  # First unload properly the module especially when the directive "remove"
+  # is used for the module
+  resource_unloading(:unload) if new_resource.reload
+
+  # Module init loading section
+  file modload_file do
+    action :delete
+  end
+
+  # Module loading options
+  resource_template_loading(:unload)
+end
+
+def modload_file
+  case node['platform_version']
+  when /^6/
+    ::File.join(node['kernel_modules']['modules_load.d'], new_resource.name + '.modules')
+  when /^7/
+    ::File.join(node['kernel_modules']['modules_load.d'], new_resource.name + '.conf')
+  end
+end
+
+def modprobe_file(action = :save)
+  # 'load' is the only special action otherwise 'save' and 'unload' should
+  # return the right path to create/remove the file
+  if action.to_s == 'load'
+    ::File.join(Chef::Config['file_cache_path'], new_resource.name + '.conf')
+  else
+    ::File.join(node['kernel_modules']['dirs']['modprobe.d'], new_resource.name + '.conf')
+  end
+end
+
+# Function to return the correct commnd for removing module
+def reload_cmd
+  # First unload the module nicely
+  unload_cmd = modprobe_bin + " -r #{new_resource.name}"
+  # If asked do it the strong way
+  unload_cmd + " || /sbin/rmmod -f #{new_resource.name}" if new_resource.force_reload
+  unload_cmd
+end
+
+def modprobe_bin(modprobe_bin = '/sbin/modprobe')
+  modprobe_bin
+end
+
+def resource_unloading(provider_action = :unload)
+  execute "Unloading module #{new_resource.name}" do
+    command reload_cmd
+    only_if "lsmod | grep -q #{new_resource.name}"
+
+    # In case of "save" or "load" we want to unload uniquely when the modules
+    # 'options' are changed only.
+    # For "unload" we unload no matter what
+    action provider_action.to_s == 'unload' ? :run : :nothing
```

Comment:
The whole logic of notifications with save, load and unload is not clear. I think that we can organize the code better